### ISSUE TITLE
Cross join large x small is much slower than small x large

### ISSUE THREAD

**Issue by thomasaarholt** [pola-rs/polars#3977]
#### What language are you using?
Python

#### What version of polars are you using?
master

#### What operating system are you using polars on?

Mac OS

#### What language version are you using
Python 3.10

#### Describe your bug.

I was writing an example dataset as a minimum working example for a stackoverflow question, and was surprised to find that a simple cross join was really slow. It turns out that if I cross join a large frame with a small frame, it takes _substantially_ longer than if I do it in the reverse order, even though the end result can be made identical via a sort.

The following `%time` is an ipython/jupyter "magic function" that times the current line once (replace it by %timeit to average several runs).

```python
import polars as pl
import numpy as np

A = pl.DataFrame({"a":np.arange(100_000)})
B = pl.DataFrame({"b":np.arange(2)})

%time A.join(B, how="cross") # Wall time: 9.61 s
%time B.join(A, how="cross") # Wall time: 274 µs
```

**Comment by thomasaarholt** [comment]
Actually, flipping the order sped up a processing step on my code from 30 minutes to 58 seconds - I had originally accepted the 30 minutes as "its big data, it takes time". ^^

**Comment by alexander-beedie** [comment]
> Actually, flipping the order sped up a processing step on my code from 30 minutes to 58 seconds - I had originally accepted the 30 minutes as "its big data, it takes time". ^^

Yikes - you're giving me flashbacks to SQLite v2, before they had any kind of optimizer, and you were entirely responsible for the order the join-loops executed. Looks like the same thing is happening here.

**Comment by ritchie46** [comment]
Is the result valid no matter the order?

**Comment by thomasaarholt** [comment]
Yes, with exception of the sorting.

**Comment by thomasaarholt** [comment]
Evidence that order doesn't matter: https://stackoverflow.com/questions/5009573/does-the-order-of-tables-in-a-join-matter-when-left-outer-joins-are-used

**Comment by ritchie46** [comment]
We can keep the order deterministic and have similar performance no matter which order. :) Fixed by #3980
